File: 3.tree_traversal/cpp/src/main.cpp
#include <iostream>

void inOrder(int* tree, int N, int idx);
void preOrder(int* tree, int N, int idx);
void postOrder(int* tree, int N, int idx);

int main(int argc, char* argv[]) {
    int N;
    int* tree;

    FILE* file = std::fopen("../test.txt", "r");
    fscanf(file, "%d", &N);
    
    tree = new int[N+1];

    for(int i=1; i<=N; i++) {
        fscanf(file, "%d", tree+i);
    }

    // in-order (left -> root -> right)
    printf("in-order: ");
    inOrder(tree, N, 1);
    printf("\n");

    // pre-order (root -> left -> right)
    printf("pre-order: ");
    preOrder(tree, N, 1);
    printf("\n");

    //post-order (left -> right -> root)
    printf("post-order: ");
    postOrder(tree, N, 1);
    printf("\n");

    return 0;
}

void inOrder(int* tree, int N, int idx) { // (left -> root -> right)
    if(idx <= N) {
        inOrder(tree, N, 2 * idx);
        printf("%d ", tree[idx]);
        inOrder(tree, N, 2 * idx + 1);
    }
}

void preOrder(int* tree, int N, int idx) { // (root-> left -> right)
    if(idx <= N) {
        printf("%d ", tree[idx]);
        preOrder(tree, N, 2 * idx);
        preOrder(tree, N, 2 * idx + 1);
    }
}

void postOrder(int* tree, int N, int idx) { // (left -> right -> root)
    if(idx <= N) {
        postOrder(tree, N, 2 * idx);
        postOrder(tree, N, 2 * idx + 1);
        printf("%d ", tree[idx]);
    }
}

File: 3.tree_traversal/python/main.py
import argparse

parser = argparse.ArgumentParser()
parser.add_argument("--input_file")
args = parser.parse_args()

def preOrder(tree, index):
    if index < len(tree):
        print("{} ".format(tree[index]), end='')
        preOrder(tree, 2*index)
        preOrder(tree, 2*index+1)

def inOrder(tree, index):
    if index < len(tree):
        inOrder(tree, 2*index)
        print("{} ".format(tree[index]), end='')
        inOrder(tree, 2*index+1)

def postOrder(tree, index):
    if index < len(tree):
        postOrder(tree, 2*index)
        postOrder(tree, 2*index+1)
        print("{} ".format(tree[index]), end='')


if __name__ == "__main__":
    with open(args.input_file, "r") as f:
        N = int(f.readline())
        tree = list(map(lambda x: int(x), f.readline().split(" ")))
        tree.insert(0, -1)

    print("Pre-Order: ", end='') 
    preOrder(tree, 1)
    print("")

    print("In-Order: ", end='') 
    inOrder(tree, 1)
    print("")

    print("Post-Order: ", end='') 
    postOrder(tree, 1)
    print("")